### PR TITLE
[alpha.webkit.UncountedLocalVarsChecker] Recursive functions are erroneously treated as non-trivial

### PR DESCRIPTION
**Pull request by rniwa**
This PR fixes the bug that alpha.webkit.UncountedLocalVarsChecker erroneously treats a trivial recursive function as non-trivial. This was caused by TrivialFunctionAnalysis::isTrivialImpl which takes a statement as an argument populating the cache with "false" while traversing the statement to determine its triviality within a recursive function in TrivialFunctionAnalysisVisitor's WithCachedResult. Because IsFunctionTrivial honors an entry in the cache, this resulted in the whole function to be treated as non-trivial.

Thankfully, TrivialFunctionAnalysisVisitor::IsFunctionTrivial already handles recursive functions correctly so this PR applies the same logic to TrivialFunctionAnalysisVisitor::WithCachedResult by sharing code between the two functions. This avoids the cache to be pre-populated with "false" while traversing statements in a recurisve function.

### CODE REVIEW

File: clang/test/Analysis/Checkers/WebKit/uncounted-local-vars.cpp [hunk 1]
```diff
@@ -289,3 +289,32 @@ void foo() {
...
+
+int TreeNode::recursiveCost() {
+  // no warnings
+  unsigned totalCost = cost;
+  for (TreeNode* node = firstChild; node; node = node->nextSibling)
+    totalCost += recursiveCost();
+  return totalCost;
+}
+
+} // namespace local_var_in_recursive_function
```
Comment (t-rasmud) [inline clang/test/Analysis/Checkers/WebKit/uncounted-local-vars.cpp:320]: Can we also have a test for a recursive function that isn't trivial? say, a recursive function calling a non-trivial method?

Comment (rniwa) [reply clang/test/Analysis/Checkers/WebKit/uncounted-local-vars.cpp:320]: https://github.com/llvm/llvm-project/blob/main/clang/test/Analysis/Checkers/WebKit/uncounted-obj-arg.cpp has a whole bunch of non-trivial recursive functions but we can add one here for alpha.webkit.UncountedLocalVarsChecker as well.

Comment (rniwa) [reply clang/test/Analysis/Checkers/WebKit/uncounted-local-vars.cpp:330]: Added a test case.

**Comment by t-rasmud** [comment]
LGTM!

**Review by t-rasmud — APPROVED** [review]

**Comment by rniwa** [comment]
Thanks for the review! Will wait for bots to catch up & merge.

**Pull request merged**